+++ /lib/netifd/proto/qmi.sh
@@ -25,6 +25,14 @@
 	proto_config_add_defaults
 }

+qmi_disconnect() {
+        # disable previous autoconnect state using the global handle
+        # do not reuse previous wds client id to prevent hangs caused by stale data
+        uqmi -s -d "$device" \
+                --stop-network 0xffffffff \
+                --autoconnect > /dev/null
+}
+
 proto_qmi_setup() {
 	local interface="$1"

@@ -96,9 +104,10 @@
 		}
 	}

+	qmi_disconnect
+
 	uqmi -s -d "$device" --set-data-format 802.3
 	uqmi -s -d "$device" --wda-set-data-format 802.3
-	uqmi -s -d "$device" --sync

 	echo "Waiting for network registration"
 	while uqmi -s -d "$device" --get-serving-system | grep '"searching"' > /dev/null; do
